libs/pika/async_cuda/include/pika/async_cuda/then_with_stream.hpp

```diff
@@ -85,268 +87,108 @@ namespace pika::cuda::experimental::detail {
         }
     }
 
-    template <typename... Ts>
-    void extend_argument_lifetimes(
-        cuda_scheduler&& sched, cuda_stream const& stream, Ts&&... ts)
+    template <typename OperationState>
+    void set_value_immediate_void(OperationState& op_state)
     {
-        if constexpr (sizeof...(Ts) > 0)
-        {
-            detail::add_event_callback(
-                [keep_alive = pika::make_tuple(PIKA_MOVE(sched),
-                     PIKA_FORWARD(Ts, ts)...)](cudaError_t status) mutable {
-                    PIKA_ASSERT(status != cudaErrorNotReady);
-                    PIKA_UNUSED(status);
-                },
-                stream.get());
-        }
-    }
-
-    template <typename R, typename... Ts>
-    void set_value_immediate_void(
-        cuda_scheduler&& sched, cuda_stream const& stream, R&& r, Ts&&... ts)
-    {
-        pika::execution::experimental::set_value(PIKA_FORWARD(R, r));
-
-        // Even though we call set_value immediately, we still extend the
-        // life time of the arguments by capturing them in a callback that
-        // is triggered when the event is ready.
-        extend_argument_lifetimes(
-            PIKA_MOVE(sched), stream, PIKA_FORWARD(Ts, ts)...);
-    }
-
-    template <typename R, typename... Ts>
-    void set_value_event_callback_void(
-        cuda_scheduler&& sched, cuda_stream const& stream, R&& r, Ts&&... ts)
-    {
-        detail::add_event_callback(
-            [r = PIKA_FORWARD(R, r),
-                keep_alive = pika::make_tuple(PIKA_MOVE(sched),
-                    PIKA_FORWARD(Ts, ts)...)](cudaError_t status) mutable {
-                set_value_event_callback_helper(status, PIKA_MOVE(r));
+        pika::visit(
+            [&op_state](auto& result) {
+                if constexpr (std::is_same_v<std::decay_t<decltype(result)>,
+                                  pika::monostate>)
+                {
+                    pika::execution::experimental::set_value(
+                        PIKA_MOVE(op_state.receiver));
+                }
+                else
+                {
+                    PIKA_UNREACHABLE;
+                }
             },
-            stream.get());
+            op_state.result);
```

Comment:
As pointed out by @aurianer this could be simplified. We can assert that the variant holds a `pika::monostate` and access that member directly.

Reply:
I've made the changes here. Please have a look @aurianer. It should be significantly more straightforward to understand now.